topdown: Protect topdown on heterogeneous processors with rseq

willowec opened the pull request:
## Pull Request Description

This PR extends #286 with [`librseq`](https://github.com/compudj/librseq) to guarantee that attempting to access the `PERF_METRICS` MSR on an Intel heterogeneous multicore processor will never result in a segmentation fault in the event of an unexpected process migration from p-core to e-core.  

The topdown component dynamically loads `librseq`, and falls back to the protections introduced in #286 if `librseq` cannot be found or loaded. This PR is tested on a Raptor Lake machine, with patched version of `librseq` that fixes an issue preventing it from being dynamically loaded. 

As of the opening of this PR, `librseq` cannot be dynamically loaded due to a static thread-local storage structure it uses. However, the maintainer has indicated that this will be removed in the future (see [`librseq` #22](https://github.com/compudj/librseq/issues/22#issuecomment-2649275555)). Until it is removed, there is no urgency to merging this PR as it doesn't offer any benefits without a dynamically loadable `librseq`.

Thanks to Mathieu Desnoyers for the help understanding how restartable sequences work, for creating `librseq`, and for the help with how to use it. 

## Author Checklist
- [x] **Description**
_Why_ this PR exists. Reference all relevant information, including _background_, _issues_, _test failures_, etc
- [x] **Commits**
_Commits_ are self contained and only do one thing
_Commits_ have a header of the form: `module: short description`
_Commits_ have a body (whenever relevant) containing a detailed description of the addressed problem and its solution
- [x] **Tests**
The PR needs to pass all the tests

jagode commented:
This is also ready to be merged.  It was also tested on a Sapphire Rapids system.

jagode commented:
@willowec : Willow, are you able to rebase and merge, please?

willowec commented:
Hm, I am getting a slightly confusing error message here. I will look into it further this afternoon

![rebase issue](https://github.com/user-attachments/assets/1249677e-78b6-468e-b31a-7ce57d28c393)

willowec commented:
@jagode I don't think I actually have permission to merge PRs?

willowec commented:
@jagode I have fixed the conflicts I believe, let me know if there are any issues. In my sanity check afterwards I caught that I had omitted clobbers on the inline assembly for the restartable sequence critical section in topdown.c, causing the component to produce invalid results when librseq was in use. I have fixed that now, and the component should be set for merge

jagode commented:
@willowec : Thanks for making these changes! I’m still seeing a message that the branch can’t be rebased due to conflicts. I’m not sure why that is happening. 

<img width="599" alt="Screenshot 2025-04-19 at 11 43 02 PM" src="https://github.com/user-attachments/assets/3016bcca-09ff-43ac-95b5-dde648c8f671" />

willowec commented:
Oh you have to be kidding me

willowec commented:
@jagode I believe this should be resolved now? I ended up hard resetting to match origin/master and then adding in the changes from there. Locally when I tried to rebase it claimed that the branch is already up to date:
![image](https://github.com/user-attachments/assets/07a4579d-813e-422a-8c96-60f61ba302c4)
But after force pushing Github still wanted me to fix some merge conflicts, so I also did that via the GitHub web GUI.

I think this leaves the PR ready to be cleanly merged, but the diffs on individual files are unfortunately pretty bad - the diff just shows that I completely deleted the old version and added in completely brand new code, making it harder to tell what changes in this PR were already present in #286.

I might be able to get a cleaner looking diff by making an entirely new branch and trying from there? I am willing to try it if you would like.
![image](https://github.com/user-attachments/assets/4801ae10-5f9b-41e8-9fe9-18d709513dc6)

Sorry about how messy this PR has been

willowec commented:
It seems like every time I fix it by rebasing locally and force pushing, if anything changes in the master branch between my fix and an attempt to merge then the rebase problem reappears

willowec commented:
Okay, I think the problem was that my development fork of papi's master branch was way out of sync. Now that I have rebased this branch onto a properly synchronized master branch, I think we are finally all set? There was no request for me to manually merge to fix conflicts in the UI this time.
![image](https://github.com/user-attachments/assets/aa4b4a27-80d6-489d-b73c-74bfc5e7f958)
